api/hasura/actions/create_user.ts

```diff
@@ -0,0 +1,116 @@
+import type { VercelRequest, VercelResponse } from '@vercel/node';
+import fetch from 'node-fetch';
+
+import { gql } from '../../../api-lib/Gql';
+import {
+  profiles_constraint,
+  ValueTypes,
+} from '../../../src/lib/gql/zeusHasuraAdmin';
+import { createUserSchemaInput } from '../../../src/lib/zod';
+
+// @ts-ignore
+global.fetch = fetch;
+
+export default async function handler(
+  request: VercelRequest,
+  response: VercelResponse
+) {
+  // TODO authenticate that the request originated from hasura
+  console.error(request.body);
+  console.error(request.body.input);
+
+  // Input Validation
+  const input: ValueTypes['createUserInput'] = request.body.input;
+  const result = await createUserSchemaInput.safeParseAsync(input);
+  // a `switch` gets typescript to behave soundly
+  // an `if` statement doesn't recognize the result
+  // disambiguation for some reason
+  console.error('result: ', result);
+  switch (result.success) {
+    case false:
+      return response.status(422).json({
+        extensions: result.error.issues,
+        message: 'invalid input',
+        code: 422,
+      });
+  }
+
+  // External Constraint Validation
+  // I'd prefer to add this uniqueness constraint into the database
+  const { object: createUserParams, circle_id } = result.data;
+
+  const { users: existingUsers } = await gql.q('query')({
+    users: [
+      {
+        limit: 1,
+        where: {
+          name: { _eq: createUserParams.name },
+          address: { _eq: createUserParams.address },
+          circle_id: { _eq: circle_id },
+          // ignore soft_deleted users
+          deleted_at: { _is_null: true },
+        },
+      },
+      {
+        id: true,
+      },
+    ],
+  });
+
+  if (existingUsers.length === 1) {
+    return response.status(422).json({
+      extensions: [],
+      message: 'User Exists',
+      code: '422',
+    });
+  }
+
+  // Update the state after all validations have passed
+  const mutationResult = await gql.q('mutation')({
+    // Insert the user
+    insert_users_one: [
+      { object: { ...createUserParams, circle_id } },
+      { id: true },
+    ],
+    // End any active nomination
+    update_nominees: [
+      {
+        _set: { ended: true },
+        where: {
+          circle_id: { _eq: circle_id },
+          address: { _eq: createUserParams.address },
+          name: { _eq: createUserParams.name },
+          ended: { _eq: false },
+        },
+      },
+      { returning: { id: true, address: true, name: true } },
+    ],
+    // Create a profile if none exists yet
+    insert_profiles_one: [
+      {
+        object: { address: createUserParams.address },
```

Comment:
FYI you can also do a nested insert instead of multiple mutations at the top level since it's for a related table to the user insert: https://hasura.io/docs/latest/graphql/core/databases/postgres/mutations/insert.html#one-to-one-one-to-many-relationships

This is fine for now too, just wanted to shout out nested inserts because they can be super useful (don't have to specify the primary key since it takes it from the foreign key inserted into the user table)

Reply:
@zashton I wonder if this is possible for circle with a user in it

Reply:
yeah possible if its a newly added user that has never login to our system, they would have no profile.